### DISCUSSION TITLE
header include order

### DISCUSSION THREAD

**Discussion by willie** [microsoft/vcpkg#36881] [Q&A] [answered]
I'm moving parts of an older project over to `vcpkg`, and it uses two pkgs, `zlib` and `cryptopp`. Unfortunately, when I try to access zlib's `zlib.h`, it's pulling in cryptopp's `zlib.h` instead and they are incompatible, because they are totally different headers.

Before moving to vcpkg fully, I managed this problem by making sure that zlib's headers were included first, but I'm not sure how to tell `vcpkg` or `cmake` how to do this.

FWIW, I have both in my vcpkg.json, and I include them in my CMakeLists.txt file like so:

```
find_package (ZLIB REQUIRED)
target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE ZLIB::ZLIB)

...

find_package(cryptopp CONFIG REQUIRED)
target_link_libraries(${CMAKE_PROJECT_NAME} PRIVATE cryptopp::cryptopp)
get_target_property(INCLUDE_DIRS cryptopp::cryptopp INTERFACE_INCLUDE_DIRECTORIES)
foreach(dir ${INCLUDE_DIRS})
	target_include_directories(${CMAKE_PROJECT_NAME} PRIVATE ${dir}/cryptopp)
endforeach()
```

Perhaps it's how I'm ensuring that `crytopp` headers are included? (If I don't do the extra part, it won't find them at all.)

**Comment by Osyotr** [comment, marked as answer]
My quick search over linux packages revealed that:
1) Some distributions install includes into `includes/cryptopp/` (notably [RHEL](https://rhel.pkgs.org/9/epel-x86_64/cryptopp-devel-8.6.0-1.el9.x86_64.rpm.html))
2) Some distributions install includes into `includes/crypto++/` (notably [Debian](https://debian.pkgs.org/12/debian-main-amd64/libcrypto++-dev_8.7.0+git220824-1_amd64.deb.html))
3) None of them provide cmake configs, but all of them provide pkg-config files
4) None of them add `cryptopp` or `crypto++` to the includes directories.

You probably should use `#include <cryptopp/*.h>` instead of `#include <*.h>` in your project.

**Reply by willie** [reply]
You are right. Our historical usage predates the nested include structure, but updating it is probably the best solution.

**Comment by willie** [comment]
If I change all my `cryptopp` header references to `cryptopp/*`, and remove the `foreach`, etc., then it works, but now a different question: how do you set header include order with cmake? Hmmm.
